UPGRADE.md

```diff
@@ -35,6 +35,152 @@ starts new workers, which take over from old workers before those old workers
+will be different depending on which previous version from which you are migrating.
+
+Upgrading into 3.0.x is a major upgrade,
+be aware of any [breaking changes](https://github.com/Kong/kong/blob/release/3.0.x/CHANGELOG.md#breaking-changes)
+between the between 2.x and 3.x series. For 1.x series, also referer to
+[breaking changes of 2.x](#breaking-changes-2.0.0).
+
+### Dependencies
+
+If you are using the provided binary packages, all necessary dependencies
+for the gateway are bundled and you can skip this section.
+
+If you are building your dependencies by hand, there are changes since the
+previous release, so you will need to rebuild them with the latest patches.
+
+The required version of OpenResty is bumped up to [1.21.4.1](https://openresty.org/en/ann-1021004001.html).
+We recommend you to use the [openresty-build-tools](https://github.com/Kong/kong-build-tools/tree/master/openresty-build-tools),
+which allows you to more easily build OpenResty with the necessary patches and modules.
+
+
+### Template changes
+
+There are **Changes in the Nginx configuration file**, between kong 2.0.x,
+2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x, 2.8.x and 3.0.x
+
+To view the configuration changes between versions, clone the
+[Kong repository](https://github.com/kong/kong) and run `git diff`
+on the configuration templates, using `-w` for greater readability.
+
+Here's how to see the differences between previous versions and 3.0.x:
+
+```
+git clone https://github.com/kong/kong
+cd kong
+git diff -w 2.0.0 3.0.0 kong/templates/nginx_kong*.lua
+```
+
+**Note:** Adjust the starting version number
+(2.0.x, 2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x) to the version number you are currently using.
+
+To produce a patch file, use the following command:
+
+```
+git diff 2.0.0 3.0.0 kong/templates/nginx_kong*.lua > kong_config_changes.diff
+```
+
+**Note:** Adjust the starting version number
+(2.0.x, 2.1.x, 2.2.x, 2.3.x, 2.4.x, 2.5.x, 2.6.x, 2.7.x) to the version number you are currently using.
+
+
+### Suggested upgrade path
+
+Follow the instructions for your backing data store to migrate to the new version.
+If you prefer to use a fresh data store and only migrate your `kong.conf` file,
+see the instructions to
+[install 3.0.x on a fresh data store](#install-30x-on-a-fresh-data-store).
+
+As with all upgrades make a backup of your data store before beginning the process.
+
+You should not configure with admin API during migration, as it may lead to unexpected behavior and
+break your configuration.
+
+**Version prerequisites for migrating to version 3.0.x**
+
+If you are migrating from 2.7.x or lower version, you should first migrate to 2.8.1.
+Here's how you can migrate to 2.8.1: [Upgrade from 1.0.x - 2.2.x to 2.8.x](#upgrade-from-10x---22x-to-28x)
+
+Once you migrated to 2.8.x, you can follow the instructions in the section
+below to migrate to 3.0.x.
+
+### Upgrade from 2.8.x (x>=2) to 3.0.x for Traditional mode
```

Comment:
```suggestion
### Upgrade from 2.8.x to 3.0.x for Traditional mode
```

That is a very strange way to describe version number, I suggest we just remove it.